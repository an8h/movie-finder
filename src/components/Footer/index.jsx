import React from 'react';
import styles from './style.css';

const Footer = () => (
  <footer>
    <div className={styles.text}>Powered by Anthoula Alipasali</div>
  </footer>
);

export default Footer;
