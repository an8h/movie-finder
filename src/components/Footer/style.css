footer {
    background-color: var(--blue-color);
    position: fixed;
    bottom: 0;
    width: 100%;
    padding: 5px 0;
  }
  
  :local(.text) {
    color: var(--white-color);
    text-align: center;
  }
  