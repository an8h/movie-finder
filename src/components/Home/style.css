:local(.container) {
    width: 80%;
    margin: 0 auto;
    overflow: hidden;
}
:local(.horizontalRow) {
    flex-direction: row;
    flex-flow: row;
    flex-wrap: nowrap;
    overflow-x: auto;
    white-space: nowrap;
    float: none;
    height: 100%;
}