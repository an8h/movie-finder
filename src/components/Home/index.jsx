/* eslint-disable react/jsx-one-expression-per-line */
/* eslint-disable no-nested-ternary */
/* eslint-disable jsx-a11y/alt-text */
/* eslint-disable react/no-array-index-key */
/* eslint-disable react/jsx-curly-newline */
/* eslint-disable react/forbid-prop-types */
/* eslint-disable camelcase */

/* 
Home Component: 
It is the first shown page when the user runs the app. 
Rendered UI:
logo                    : The logo image of the app
SearchBar               : The serach bar of the app an a Search Button
title                   : The title of each shown category
Slider                  : A Slider component for each category. When an item is clicked the DetailPage opens with the coresponding item
div (Text "Loading...") : It is shown while the data are being fetched
div --> p               : It is shown when there is an error on API calls
*/

import React from 'react';
import { Redirect } from 'react-router-dom';
import { PropTypes } from 'prop-types';
import { connect } from 'react-redux';
import { mapDispachToProps } from '../../redux/Store';
import SearchBar from '../SearchBar/index';
import Slider from '../Slider/index';
import styles from './style.css';
import Footer from '../Footer/index';
import logo from '../../img/logo.svg';
import { POSTER_SIZE, FALLBACK_IMAGE_URL } from '../../config';
import {
  getCorrectPosterSize,
  getPosterPath,
  getTitle,
} from '../../utils/utils';

class Home extends React.Component {
  constructor() {
    super();
    this.state = { clicked: false };

    // Initializing reference to click handler
    this.handleClick = this.handleClick.bind(this);
  }

  componentDidMount() {
    const {
      getConfig,
      getGenres,
      getPopularMovies,
      getPopularSeries,
    } = this.props;

    // Make the API calls and get the needed data
    getConfig();
    getGenres();
    getPopularMovies();
    getPopularSeries();
  }

  // Keep track of the clicked item. We need to pass these properties to DetailPage Component
  handleClick(item, posterToShow) {
    this.setState({
      clicked: true,
      clickedItem: item,
      posterToShow,
    });
  }

  render() {
    const { config } = this.props;
    const { base_url = FALLBACK_IMAGE_URL } = config.images;
    const { poster_sizes } = config.images;
    const { popularMovies } = this.props;
    const popularMoviesData = popularMovies.data;
    const { popularSeries } = this.props;
    const popularSeriesData = popularSeries.data;
    const { genres } = this.props;
    const familyGenre = genres.family;
    const familyGenreData = familyGenre.data;
    const documentaryGenre = genres.documentary;
    const documentaryGenreData = documentaryGenre.data;
    const { clickedItem } = this.state;
    const { clicked } = this.state;
    const { posterToShow } = this.state;

    const correctPosterSize = getCorrectPosterSize(poster_sizes);
    const posterSize = correctPosterSize
      ? correctPosterSize[0]
      : POSTER_SIZE;

    const moviesPosterPath = getPosterPath(
      popularMovies,
      base_url,
      posterSize,
    );
    const moviesTitle = getTitle(popularMovies);

    const seriesPosterPath = getPosterPath(
      popularSeries,
      base_url,
      posterSize,
    );
    const seriesTitle = getTitle(popularSeries);

    const familyPosterPath = getPosterPath(
      familyGenre,
      base_url,
      posterSize,
    );
    const familyTitle = getTitle(familyGenre);

    const documentaryPosterPath = getPosterPath(
      documentaryGenre,
      base_url,
      posterSize,
    );
    const documentaryTitle = getTitle(documentaryGenre);

    /* The following arrays are needed for the render(). 
    In the render() we loop through the different categories and we create the UI */
    const posterPaths = [];
    posterPaths.push(
      moviesPosterPath,
      seriesPosterPath,
      familyPosterPath,
      documentaryPosterPath,
    );

    const titles = [];
    titles.push(
      moviesTitle,
      seriesTitle,
      familyTitle,
      documentaryTitle,
    );

    const categoryTitles = [
      'Popular Movies',
      'Popular Series',
      'Family',
      'Documentary',
    ];

    const data = [];
    data.push(
      popularMoviesData,
      popularSeriesData,
      familyGenreData,
      documentaryGenreData,
    );

    const dataAreLoading = [
      popularMovies.isLoading,
      popularSeries.isLoading,
      genres.isLoading,
      genres.isLoading,
    ];

    const dataLoadingNotSuccessful = [
      popularMovies.errors,
      popularSeries.errors,
      genres.errors,
      genres.errors,
    ];

    return (
      <div>
        <div className={styles.container}>
          <div>
            <img className={styles.logo} src={logo} />
          </div>
          <SearchBar base_url={base_url} posterSize={posterSize} />
          {categoryTitles.map((title, index) => (
            <div key={index}>
              <h1 className={styles.categoryTitle}>{title}</h1>
              {dataAreLoading[index] === false &&
              dataLoadingNotSuccessful[index].length <= 0 ? (
                <div className={styles.horizontalRow}>
                  {posterPaths[index].map((posterPath, key) => (
                    <Slider
                      key={key}
                      posterToShow={posterPath}
                      title={titles[index][key]}
                      onClick={() =>
                        this.handleClick(data[index][key], posterPath)
                      }
                    />
                  ))}
                </div>
              ) : dataLoadingNotSuccessful[index].length <= 0 ? (
                <div>Loading...</div>
              ) : (
                <div>
                  <p>
                    Something went wrong. Could not get {title} data.
                    Please try again.
                  </p>
                </div>
              )}
            </div>
          ))}
          {clicked ? (
            <Redirect
              push
              to={{
                pathname: '/detail',
                state: { clickedItem, posterToShow, clicked },
              }}
            />
          ) : null}
        </div>
        <Footer />
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    ...ownProps,
    config: state.config,
    genres: state.genres,
    popularMovies: state.popularMovies,
    popularSeries: state.popularSeries,
  };
};

Home.propTypes = {
  getConfig: PropTypes.func.isRequired,
  getGenres: PropTypes.func.isRequired,
  getPopularMovies: PropTypes.func.isRequired,
  getPopularSeries: PropTypes.func.isRequired,
  popularMovies: PropTypes.object.isRequired,
  popularSeries: PropTypes.object.isRequired,
  genres: PropTypes.object.isRequired,
  config: PropTypes.object.isRequired,
};

export default connect(mapStateToProps, mapDispachToProps)(Home);
