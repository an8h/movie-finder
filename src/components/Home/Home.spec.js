import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import Home from './index';
import Footer from '../Footer/index';

const mockStore = configureStore([]);

describe('Home Component', () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({
      popularMovies: {
        totalPages: 1,
        data: [],
        isLoading: false,
        errors: [],
      },
      popularSeries: {
        totalPages: 1,
        data: [],
        isLoading: false,
        errors: [],
      },
      genres: {
        family: {
          totalPages: 1,
          data: [],
        },
        documentary: {
          totalPages: 1,
          data: [],
        },
        isLoading: false,
        errors: [],
      },
      config: {
        images: {},
        isLoading: false,
        errors: [],
      },
    });

    store.dispatch = jest.fn();

    component = renderer.create(
      <Provider store={store}>
        <Home
          popularMovies={store.popularMovies}
          popularSeries={store.popularSeries}
          genres={store.genres}
          config={store.config}
          familyGenreData={store.genres}
        />
      </Provider>,
    );
  });

  it('should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});
