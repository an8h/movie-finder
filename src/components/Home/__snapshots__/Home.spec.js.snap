// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component should render with given state from Redux store 1`] = `
<div>
  <div
    className="container"
  >
    <div>
      <img
        className="logo"
        src={
          Object {
            "process": [Function],
          }
        }
      />
    </div>
    <form
      onSubmit={[Function]}
    >
      <div
        className="container"
      >
        <input
          className="inputCustom"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="text"
          value="Search for...."
        />
        <button
          className="buttonCustom"
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
    <div>
      <h1
        className="categoryTitle"
      >
        Popular Movies
      </h1>
      <div
        className="horizontalRow"
      />
    </div>
    <div>
      <h1
        className="categoryTitle"
      >
        Popular Series
      </h1>
      <div
        className="horizontalRow"
      />
    </div>
    <div>
      <h1
        className="categoryTitle"
      >
        Family
      </h1>
      <div
        className="horizontalRow"
      />
    </div>
    <div>
      <h1
        className="categoryTitle"
      >
        Documentary
      </h1>
      <div
        className="horizontalRow"
      />
    </div>
  </div>
  <footer>
    <div
      className="text"
    >
      Powered by Anthoula Alipasali
    </div>
  </footer>
</div>
`;
